### ISSUE TITLE
Next-tick and periodic callbacks always block the thread, even with non-blocking code

### ISSUE THREAD

**Issue by jdugge** [bokeh/bokeh#3991]
It seems that any callback added using `add_next_tick_callback` or `add_periodic_callback` will block the thread, even if it's a non-blocking `@gen.coroutine`.

For instance, I would expect the provided minimal example to print `periodic callback` about ten times a second, and `next tick callback` roughly once a second. However, `periodic callback` is never printed: the next tick callback blocks the thread so the periodic callback isn't executed.

Is this the expected behaviour? If it is, how would I correctly add a non-blocking callback? If I'm reading it right, the documentation (http://bokeh.pydata.org/en/0.11.0/docs/dev_guide/server.html#tornado-ioloop-and-async-code) suggests that non-blocking `@gen.coroutine`s in callbacks are the way to go, but maybe I'm doing it wrong?

```python
import bokeh.plotting
from tornado import gen

@gen.coroutine
def my_non_blocking_function():
    print "next tick callback"
    
    # Non-blocking sleep
    yield gen.sleep(1)
    
    # Add callback again
    doc.add_next_tick_callback(my_non_blocking_function)

def my_periodic_function():
    print "periodic callback"

doc = bokeh.plotting.curdoc()

# Call my_periodic_function every 100 ms
doc.add_periodic_callback(my_periodic_function, 100)

# Add my_non_blocking_function
doc.add_next_tick_callback(my_non_blocking_function)
doc.add_root(p)
```

**Comment by havocp** [comment]
This seems like a bug to me, though I could be missing some detail. Nonblocking callbacks should block the next invocation of themselves (i.e. if you add a periodic callback that yields, it shouldn't run again until it finishes), but separately-added callbacks should not be affecting one another.

**Comment by jdugge** [comment]
As @bryevdv points out in #3702, the example works as expected when the callback is added using `tornado.ioloop.IOLoop.instance().add_callback(my_non_blocking_function)` instead of using `bokeh.plotting.curdoc().add_next_tick_callback(my_non_blocking_function)`:

```python
import bokeh.plotting
from tornado import gen
import tornado

@gen.coroutine
def my_non_blocking_function():
    print("next tick callback")

    # Non-blocking sleep
    yield gen.sleep(1)

    # Add callback again
    tornado.ioloop.IOLoop.instance().add_callback(my_non_blocking_function)

def my_periodic_function():
    print("periodic callback")

doc = bokeh.plotting.curdoc()

# Call my_periodic_function every 100 ms
doc.add_periodic_callback(my_periodic_function, 100)

# Add my_non_blocking_function
doc.add_next_tick_callback(my_non_blocking_function)
doc.add_root(p)
```

**Comment by bryevdv** [comment]
I believe this is a dupe of #3702 which has just been closed by #4152
